cmake/CeleritasUtils.cmake

```diff
@@ -73,13 +85,66 @@ CMake utility functions for Celeritas.
     [<INTERFACE|PUBLIC|PRIVATE> [items2...] ...])
 
 #]=======================================================================]
-include(FindPackageHandleStandardArgs)
+include_guard(GLOBAL)
 
-macro(celeritas_find_package_config _package)
-  find_package(${_package} CONFIG ${ARGN})
-  find_package_handle_standard_args(${_package} CONFIG_MODE)
+include(CheckLanguage)
+
+#-----------------------------------------------------------------------------#
+# CONFIGURATION UTILITIES
+#-----------------------------------------------------------------------------#
+
+function(celeritas_optional_language lang)
+  set(_var "CELERITAS_USE_${lang}")
+  if(DEFINED "${_var}")
+    set(_val "${_var}")
+  else()
+    check_language(${lang})
+    set(_val OFF)
+    if(CMAKE_${lang}_COMPILER)
+      set(_val ON)
+    endif()
+    message(STATUS "Set ${_var}=${_val} based on compiler availability")
+  endif()
+
+  option("${_var}" "Enable the ${lang} language" "${_val}" )
+endfunction()
+
+function(celeritas_to_onoff varname)
+  if(ARGC GREATER 1 AND ARGV1)
+    set(${varname} ON PARENT_SCOPE)
+  else()
+    set(${varname} OFF PARENT_SCOPE)
+  endif()
+endfunction()
+
+# Note: this is a macro so that `find_package` variables stay in the global
+# scope.
+macro(celeritas_optional_package package)
```

Comment:
Should this routine support the `COMPONENTS` parameter (of `find_package`), in particular this could be use to require a CUDA enable build of VecGeom.

Reply:
Nevermind :), this does not actually do the `find_package` ...

Reply:
It doesn't have to, no. This setup step will only ever happen on the first configure (or if the user deletes the cmake variable) and just needs to test for the existence of the package. The `find_package(Foo ...)` later on in the main CMakeLists is what enforces the versions and components and such.

It would *ideally* be cool to have the extra components/version restrictions as an option, since currently (for example) having SWIG 3.0 on your system without setting `CELERITAS_USE_SWIG=OFF` results in a configure failure, since it detects SWIG (setting the variable to ON by default) but then later in the code fails the version check. However, such a feature would require extra complexity in this macro, make the `celeritas_optional_package` chunk of code more difficult to read, and effectively duplicate the further-down `find_package` requirement. So I don't think those tradeoffs are worth it. (My solution to the SWIG issue was to disable it by default in the cmake preset file.)

Reply:
What would be the downside if you combined the two calls.  I.e. where you have the `find_package` we would have
```
celeritas_declare_option_and_find_package(nlohmann_json 3.7.0)
```
without improvement this would currently be:
```
macro(celeritas_declare_option_and_find_package package)
   celeritas_optional_package(package)
   set(_var "CELERITAS_USE_${package}")
   if (${_var}) 
       find_package(${package} ${ARGN} REQUIRED)
   endif()
   undef(_var)
endmacro()
```

Reply:
The downsides I see are:
- Added complexity for the macro and less easily understandable cmake code: before it was "create and option and set an intelligent default", now it does one more thing; and it also makes it less obvious where the `find_package` variables are coming from and what their scope is
- The `option` no longer fits on one line so it's harder to read the series of options we create
- It can't handle more complex logic like the changes for the Python components required by having SWIG enabled

Reply:
I think the key thing is that the code inside `celeritas_optional_package` only affects anything on the initial configure (and even then simply defines an option and a default value). Changing it to also *include* the dependency means it becomes critical to the rest of the CMake logic: keeping the `find_package` explicit at the top level makes that logic more transparent.

Reply:
Does 'find_package' re-run (and thus can find a different answer) at each cmake configure invocation?

Reply:
> (and even then simply defines an option and a default value).

And also runs `find_package` and set all the related variable (for some 'arbitrary' version of the package) ... [This ought to be mostly innocuous but some older version of the cmake fragment might (inadvertently) set/use 'bad' variable names]

Reply:
i.e I still find it cumbersome that (at the first config), find_package is run twice.

Reply:
> It can't handle more complex logic like the changes for the Python components required by having SWIG enabled

why not?

Reply:
If I understood correctly, the pattern this code allows is:

* If a package is available, default to making it a requirement (which the right version or settings).

Another alternative implementation might:

* default all optional to "Try" (i.e allowed value ON/OFF/TRY
* have a find_package_wrapper that does:
```
if (TRY)
   find_package(packagename ${options} QUIET)
   if (FAILED)
      find_package(packagename QUIET)  # See if there is any other packages
      if (SUCCESS)
          find_package(packagename ${options})   # Finally get the error message
else if (ON)
   find_package(packagename ${options})   # Finally get the error message
```

Hummm ... right ... you are going to **correctly** point out that this is too complicated.

So what about a 4th options :)

* default all optional to "Try" (i.e allowed value ON/OFF/TRY
* have a find_package_wrapper that does:
```
if (TRY)
   find_package(packagename ${options} QUIET)
   if (FAIL)
		set option to OFF
else if (ON)
   find_package(packagename ${options})   
```

Reply:
@pcanal `find_package` re-runs but typically uses existing cache variables to shortcut some of the `find` logic. The real problem is that CMake is just so fragile, and each find module can behave very differently.

Yes it's a cumbersome to run `find_package` twice at the first configure, but generally speaking the second `find_package` should be much faster and not break anything, and I think running `find_package` an extra time on the initial configure is less cumbersome than requiring the user to figure out the right options in advance, and I think it's preferable to combining both `find_package` calls into one.

I continue to think of this new macro as a convenience function for setting a default, not an integral part of the overall cmake code (which ir would be if we combined the `find_package` calls).

Reply:
> Hummm ... right ... you are going to correctly point out that this is too complicated
😂

Reply:
I think I'm also fixated on having the "default option" take up no more space than the equivalent (but dumb default) `option(FOO "Foo" OFF)`. Any change to make the macro more complicated means interleaving the option declarations with additional cmake logic for determining the defaults, and even reordering the option declarations based on arbitrarily complex dependencies (e.g. having to put SWIG before Python).

Reply:
I still think that the risk involved in running `find_package` twice with different argument is not worth it (as you said "CMake is just so fragile") and would pick the option with 'try/on/off + small find_package_wrapper'  [or even default to all on and force those without some option to disable them].

However, I also essentially agree to all your points and the real (long term) cost difference between the options is really hard to quantify and thus don't feel too strongly about which to choose (hence I clicked the "approve" button)